File: Tut_07/Java/Queue_LinkedList.java
package Tut_07;

public class Queue_LinkedList<T> {
	Node <T> front = null;
	Node <T> rear = null;
	
	// print the item
	public void printList() {
		Node temp = this.front;
		while (temp != null) {
			System.out.print(temp.data + " ");
			temp = temp.next;
		}
	}

	// isEmpty(): Check if the stack is empty
	public boolean isEmpty() {
		if ( front == rear)
			return true;
		return false;
	}
	
	public void enqueue(int data) {
		Node temp = new Node(data);
		
		if ( this.rear == null) {
			this.front = this.rear = temp;
		}
		
		// Add the new node at the end of queue and chagne rear
		this.rear.next = temp;
		this.rear = temp;
	}
	
	public void dequeue() {
		if ( this.front == null) {
			System.out.println("Nothing left to dequeue");
		}
		// Remove the first node
		this.front = this.front.next;
		
		// If front = null, -> rear = null
		if ( this.front == null) {
			this.rear = null;
		}
	}
	
	public int peek(){
		if ( this.isEmpty())
			return -1;
		return (int)this.front.data;
	}

}

File: Tut_07/Python/queue_linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    def __init__(self):
        # Node head
        self.front = None
        self.rear = None

    def isEmpty(self):
        return self.front == None

    def enQueue(self, data):
        temp = Node(data)
        if self.rear == None:
            self.front = self.rear = temp
            return
        self.rear.next = temp
        self.rear = temp

    def deQueue(self):
        if self.isEmpty():
            return 

        temp = self.front
        self.front = temp.next

        if self.front == None:
            self.rear = None

if __name__== '__main__': 
    q = Queue() 
    q.enQueue(10) 
    q.enQueue(20) 
    #q.DeQueue() 
    #q.DeQueue() 
    q.enQueue(30) 
    q.enQueue(40) 
    q.enQueue(50)  
    q.deQueue()    
    print("Queue Front " + str(q.front.data)) 
    print("Queue Rear " + str(q.rear.data)) 





